### DISCUSSION TITLE
[Lua] uosc 多功能控制界面

### DISCUSSION THREAD

**Discussion by hooke007** [hooke007/mpv_PlayKit#186] [INFO 信息]
极简主义设计驱动，支持轻度自定义，功能丰富的界面类 (osc-like) 脚本

>[!Note]
>本仓库的当前存档为基于上游 5.12.0+ 的修改版
>（有较大差异，不保证对基于原版uosc的扩展脚本的完全兼容性）

# 安装部署

1. （必选）下载此**文件夹** [**scripts/uosc**](https://github.com/hooke007/mpv_PlayKit/blob/main/portable_config/scripts/uosc/) 到你的本地脚本文件夹内
2. （必选）下载 [**fonts/MaterialIconsRound-Regular.otf**](https://github.com/hooke007/mpv_PlayKit/blob/main/portable_config/fonts/MaterialIconsRound-Regular.otf) & [**fonts/uosc_textures.ttf**](https://github.com/hooke007/mpv_PlayKit/blob/main/portable_config/fonts/uosc_textures.ttf) 到你的本地mpv字体文件夹
3. （必选）下载 [**ziggy-windows.exe**](https://github.com/hooke007/dotfiles/releases/tag/mpv_addones) 到你的本地 `uosc/bin/` 路径内
4. （可选）下载 **script-opts/uosc.conf** 到你的本地脚本设置文件夹并修改内部参数适配
5. （可选）下载 **scripts/thumbfast.lua** 到你的本地脚本文件夹（此项为缩略图显示支持）
6. （可选）下载 **script-opts/thumbfast.conf** 到你的本地脚本设置文件夹并修改内部参数适配
7. （可选）在你的本地 **input.conf** 中写入 
```
MBTN_RIGHT   script-binding uosc/menu   # 鼠标右键唤起uosc的OSD上下文菜单
```
8. （可选）修改你的本地 **input.conf** 以进一步适配uosc（此步骤的具体内容参见下文）

# 功能简介

当前使用版本的原仓库英文文档 https://github.com/tomasklaen/uosc/blob/5.12.0/README.md

## 1. 可客制化上下文菜单

即便你的 input.conf 为空，也可以使用内置的预设菜单。
如果你想要使用自定义，参见下文或 lite 分支的 [input_uosc.conf](https://github.com/hooke007/mpv_PlayKit/blob/lite/portable_config/input_uosc.conf)
（主分支的 [input_uosc.conf](https://github.com/hooke007/mpv_PlayKit/blob/main/portable_config/input_uosc.conf) 只记录预设菜单的对应语法）

语法简要说明与示例：
_只需要基于 input.conf 的常规语法，做出轻度改变。_

- 不要在行末使用注释， 用 `#!` 或 `#menu:` 作为映射菜单条目的起始标志。
- 支持绑定到同指令的不同键位映射同一菜单条目。
- 支持不绑定任何键位直接映射菜单条目。
- 可以映射指令到次级菜单。
```
f        cycle fullscreen   #! 切换 全屏
CTRL+f   cycle fullscreen   #! 切换 全屏
#        quit               #! 退出程序
#        cycle mute         #! 音量 > 切换 静音
```
以上示例汇总的预览：
<img src="https://user-images.githubusercontent.com/41094733/192123671-cce7067e-9990-424f-ab40-8ae7dd895e14.jpg" width=500>

🔁 P.S. 也可使用键盘 <kbd>UP</kbd>/<kbd>DOWN</kbd>/<kbd>LEFT</kbd>/<kbd>RIGHT</kbd> 在菜单中预选条目，<kbd>Enter</kbd> 即模拟点击条目。
🔁 P.S. 按住 <kbd>Alt</kbd> 键再点击菜单的条目，可重复触发而不会立即关闭菜单。

### 1.1. 菜单输入与搜索

_支持输入中文拼音搜索，支持从剪贴板粘贴文本。_

<img src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/fdb03399-b6dc-48a2-9fa0-a3323c27389d" width=600>

## 2. 便利的界面控件

- 速度、音量滑块可进行快速调节
- 音频设备、视频/音频/字幕轨等的可视化切换
- 控制栏元素的自定义

🔁 P.S. 在播放列表界面，对指定条目按 <kbd>Ctrl</kbd>+<kbd>UP</kbd>/<kbd>DOWN</kbd> 可移动条目，按 <kbd>DEL</kbd> 则移除条目。

<img src="https://user-images.githubusercontent.com/41094733/192123792-72e96027-1569-434e-9804-f34abf445708.jpg" width=200> <img src="https://user-images.githubusercontent.com/41094733/192123858-d5e42c0e-db13-40f4-aabc-3ff9c77f4338.jpg" height=180> <img src="https://user-images.githubusercontent.com/41094733/192123927-35874a17-20f5-4909-ab6b-6eec7393b24c.jpg" width=400>

## 3. 丰富的用户选项

其它简易选项直接参见 [uosc.conf](https://github.com/hooke007/mpv_PlayKit/blob/main/portable_config/script-opts/uosc.conf)

### 3.1. 脚本选项的扩展说明

部分用户选项较为复杂，此处针对如下选项进行详细的进一步说明。
- [x] controls
- [x] chapter_ranges
- [x] chapter_range_patterns

<hr />

#### 3.1.1. controls

```ini
controls=menu,ST-stats_tog,gap,play_pause,gap,subtitles,audio,<has_chapter>chapters,<has_many_edition>editions,<has_many_video>video,<stream>stream-quality,gap,space,speed,space,shuffle,loop-playlist,loop-file,gap,prev,items,next,gap,fullscreen
```
自定义时间线上方的控件按钮，多个值以半角逗号分隔。值 `never` 即禁用，上方的示例即默认值。

完整的条目语法（参数说明：`{值}`, `[可选]`）：
```
[<[!]{disposition1}[+[!]{dispositionN}][,{more_dispositions}]>]{element}[:{paramN}][#{badge}[>{limit}]][?{tooltip}]
```

常见属性：
`{icon}` 指定图标名称的参数。在这里查询所有可用的值（仅限Rounded类型） https://fonts.google.com/icons?selected=Material+Icons&icon.style=Rounded
👉示例：
点击图标，右侧会弹出显示你实际要用的的参数名。此处 "Dark Mode" 应写为 `dark_mode`
<img width=500 src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/eb80b77b-8ade-435d-9c2b-2ae01a860306">

`{element}` 的参数:

| 快捷指令<br>`{usoc_command}`（可直接作为按钮） | 对应功能说明 |
| -- | -- |
| `speed[:{scale}]` | 速度滑块 其中控件系数的尺寸，默认 1.3 |
| `command:{icon}:{command}` | 按下该按钮时执行的指令 |
| `toggle:{icon}:{prop}[@{owner}]` | 切换mpv属性的按钮 |
| `cycle:{default_icon}:{prop}[@{owner}]:{value1}[={icon1}][!]/{valueN}[={iconN}][!]` | 在不同mpv属性的值之间循环的按钮，每个值都可以选择不同的 `{icon}` 和激活标记。<br>结尾处赋予可选的半角感叹号，将使该按钮成为可激活的样式。<br>`{owner}` 是管理该属性的脚本名称（如果存在） |
| `gap[:{scale}]` | 留出一个间隔 其中系数的尺寸，默认 0.3 |
| `space` | 填补上两个控件之间的所有空间，对于将条目右对齐非常有用。<br>用多个 `space` 可在它们之间分配空间，可用于居中对齐 |
| `button:{name}` | 可由外部脚本改变的按钮图标的状态、外观和点击操作 |

控件条目的可见性控制：
`<[!]{disposition1}[+[!]{dispositionN}][,{more_dispositions}]>` 可选的前缀是用于控制 `{element}` 的可见性；
`+` 表示条件“和”， `,` 表示“或”。实力： `<foo,bar+baz>` -> `foo OR (bar AND baz)`

| `{disposition}` 的可用值 |  |
| -- | -- |
| `idle`             | 如果mpv处于空闲状态则为 true |
| `image`            | 如果当前文件为单帧图片则为 true |
| `audio`            | 如果当前文件为纯音频则为 true |
| `video`            | 如果当前文件存在视频轨则为 true |
| `has_many_video`   | 如果当前文件存在多个视频轨则为 true |
| `has_image`        | 如果当前文件的某个视频轨道是图片，则为 true |
| `has_audio`        | 如果当前文件存在音轨则为 true |
| `has_many_audio`   | 如果当前文件存在多个音轨则为 true |
| `has_sub`          | 如果当前文件存在字幕轨则为 true |
| `has_many_sub`     | 如果当前文件存在多个字幕轨则为 true |
| `has_many_edition` | 如果当前mkv存在多个版本则为 true |
| `has_chapter`      | 如果当前文件存在章节则为 true |
| `stream`           | 如果当前文件为流则为 true |
| `has_playlist`     | 如果当前列表存在多个文件则为 true |
| `{mpv_prop}`       | 用于判定任意mpv属性 |

_可选的前缀可用来反转所需的处理方式_

示例：
`<stream>stream-quality` 仅对流媒体显示“流式传输品质偏好”按钮
`<has_audio+!audio>audio` 对所有存在音轨的文件显示“音轨列表”按钮，但不包括纯音频的文件

将 `#{badge}[>{limit}]` 放在 `{element}` 参数后，可赋予它一个徽章标记。可用的 `badge` 值：
`sub audio video` （轨道计数值）
`{mpv_prop}` 如果mpv的属性值是一个数组，将显示其大小。可用的属性参见 https://mpv.io/manual/master/#property-list
`>{limit}` 只有当它的数值高于这个阈值时，才会显示徽章标记。
  示例： `#audio>1`

将 `?{tooltip}` 放在 `{element}` 的设置后面，赋予它一个工具提示。
  示例： `<stream>stream-quality?Stream quality`

✨_如果你好奇这个选项的默认值里的短指令是怎么实现的，这里有部分示范：_
`menu` → `command:menu:script-binding uosc/menu-blurred?Menu`
`subtitles` → `command:subtitles:script-binding uosc/subtitles#sub?Subtitles`
`fullscreen` → `cycle:crop_free:fullscreen:no/yes=fullscreen_exit!?Fullscreen`
`loop-playlist` → `cycle:repeat:loop-playlist:no/inf!?Loop playlist`
`toggle:{icon}:{prop}` → `cycle:{icon}:{prop}:no/yes!`

#### 3.1.1.1. 内置捷径汇总

除需要用户手动定制按钮外，下方的列表还列出了脚本内预设的多个预配置好的按钮，你只需要直接粘贴这些名称代号即可快速使用（`controls`的默认值就用到了其中一部分）。

|  | 功能简介 |
| -- | -- |
| `menu` | 上下文菜单 |
| `subtitles` `audio` `video` | 字幕/音频/视频轨道的查看与挂载 |
| `audio-device` | 音频设备列表 |
| `playlist` `chapters` `editions` | 播放列表 章节列表 mkv的版本列表 |
| `stream-quality` | 流式传输质量偏好 |
| `open-file` | 文件浏览加载器 |
| `items` | 播放列表（如果列表文件大于2）或<br>文件浏览加载器（如果列表文件小于2） |
| `prev` `next` | 列表中的上/下个文件（如果列表文件大于2）或<br>当前目录中的上/下个文件（如果列表文件小于2） |
| `first` `last` | 列表中的首/末位文件（如果列表文件大于2）或<br>当前目录中的首/末位文件（如果列表文件小于2） |
| `loop-playlist` `loop-file` | 列表循环 单曲循环 |
| `shuffle` | 乱序播放 |
| `autoload` | 自动播放当前路径内的下一个文件 |
| `fullscreen` | 切换全屏 |
|  |  |
| `play_pause` | 播放/暂停 |
| `pause_play` | 播放/暂停（图标反转） |
| `pl-prev` `pl-next` | 列表中的上/下个文件 |
| `border` | 开/关窗口边框 |
| `ontop` | 开/关置顶 |
| `hwdec` | 软解/硬解/增强硬解优先 |
| `unscaled` | 保持原始尺寸/缩放到目标分辨率 |
| `deband` | 开/关去色带 |
| `deint` | 开/关反交错 |
| `shot-vid` | 截屏（video模式） |
| `ST-stats_tog` | 常驻显示/隐藏统计数据 |
| `ST-thumb_tog` | 启用/禁用缩略图预览 |

<hr />

#### 3.1.2. chapter_ranges

```ini
chapter_ranges=openings:30ABF964,endings:30ABF964,ads:C54E4E80
```
将一些常见的章节类型转换成章节范围指示标记。示例即默认值
在时间轴的顶部，不再以钻石图标的形式显示章节的开始，而是根据下方的设置，对该章节所拥有的时间轴部分进行着色。

语法是一个以半角逗号分隔的列表，其中包括 `{type}:{color}`

| `{type}` 范围类型<br>当前支持的值： | 说明 |
| -- | -- |
| `openings` `endings` | 动画的 OP ED |
| `intros` `outros` | 视频的 序章 尾声 |
| `ads` | 需要此脚本支持的赞助广告<br>https://github.com/po5/mpv_sponsorblock |

`{color}` 颜色代码，格式为 `RRGGBB(AA)`

若不想要转换某个范围类型，只需从列表中移除其对应项

<hr />

#### 3.1.3. chapter_range_patterns

```ini
chapter_range_patterns=openings:オープニング;endings:エンディング
```
补充额外的lua模式来识别简单章节范围的起始点（除 `ads` 外的所有章节）。示例即默认值
语法说明：
`{type}:{pattern}[,{patternN}][;{type}:{pattern}[,{patternN}]]`

<hr />

## 4. 支持缩略图引擎（非必要）

<img src="https://user-images.githubusercontent.com/41094733/192124073-5bb7287d-bb82-4785-b846-a3ecacbb80b3.jpg" width=300>

需要额外脚本 [thumb_engine](https://github.com/hooke007/mpv_PlayKit/discussions/668) 共同实现

## 5. 其它可选的功能扩展/附属脚本

- https://github.com/hooke007/mpv_PlayKit/discussions/669
- https://github.com/christoph-heinrich/mpv-quality-menu
- https://github.com/po5/mpv_sponsorblock
- ...

# 额外注意事项

由于此脚本功能复杂，与多个脚本存在功能重合/冲突，自行留意并解决潜在的问题（禁用部分功能或<s>删除文件</s>）。

详见 [《维基_外置脚本路径》](https://github.com/hooke007/mpv_PlayKit/wiki/2_portable_config#%E5%A4%96%E7%BD%AE%E8%84%9A%E6%9C%AC%E8%B7%AF%E5%BE%84-)

**Comment by zhongfly** [comment]
补充：
一个“播放/暂停”控件：
```
cycle:play_arrow:pause:no=pause/yes=play_arrow?播放/暂停
```
暂停时：
![image](https://user-images.githubusercontent.com/11155705/192144112-4359596e-1114-47fc-bc9c-775862fccdd9.png)
播放时：
![image](https://user-images.githubusercontent.com/11155705/192144128-0dbcf2ff-4ef9-42ad-ad0b-2b04f91168a4.png)

上一章节：
```
<has_chapter>command:skip_previous:add chapter -1?上一章节
```
下一章节：
```
<has_chapter>command:skip_next:add chapter 1?下一章节
```

综合使用这三个控件：
![image](https://user-images.githubusercontent.com/11155705/192148183-6b4117bd-77db-46d0-8cb9-6fe81229ce1c.png)

显示stats：`command:analytics:script-binding stats/display-stats-toggle?状态`

~~奇怪的是，我在mpv.net里使用时，有章节的视频会变得全白，不显示ui，右键失效，log一切正常，没有任何提示，什么原因？（完全相同portable_config的mpv.exe正常显示）~~

**Reply by hooke007** [reply]
> 不显示ui 右键失效

尝试解绑 mpvnet 原右键菜单，追加主设置 `--osc=no` (此外不清楚mpvnet的UI图标会不会冲突)
如仍然无法解决需要向 mpvnet 项目反馈

edit: 顺带一提mpvnet屏蔽了画面中间的绝大块可交互区域（只允许上下两端的一小部分），目的是为了触发它的右键菜单不与原始osc交互冲突

**Reply by zhongfly** [reply]
无章节的视频使用正常，并且右键还是mpv.net自己的菜单；但是打开有章节的，就出问题了，很神奇

**Reply by zhongfly** [reply]
```
chapter_ranges=
chapter_range_patterns=
```
关闭usoc这里关于章节的设置,然后删除watch_later里对应文件的播放位置记录，似乎就正常了

正常之后再加回这些设置，还是能正常播放，搞不明白了

**Reply by dyphire** [reply]
uosc中间放控件的最大问题就是如果左右控件是按条件显隐的话就有不小的概率会不居中，这点默认的中间速度控件也会遇到。
ui控件动态显示的问题就是这个了，比起显隐控件或许明暗度区分状态对uosc这种ui绘制方式来说更合理

**Comment by OiCkilL** [comment]
我想只使用 uosc 的菜单和 [mpv-osc-modern](https://github.com/maoiscat/mpv-osc-modern) 的话，要怎么配置啊？

**Reply by zhongfly** [reply]
想要右键菜单的话，可以用mpv.net这个GUI 可以与osc脚本共存

要不就你自己修改脚本，给那个osc加上菜单

**Reply by hooke007** [reply]
已同步上游的关联功能。
https://github.com/hooke007/mpv_PlayKit/blob/ed8c27b918e56de3176514e90dcb7f4beb0e8790/portable_config/script-opts/uosc.conf#L143-L145

**Comment by txyyh** [comment]
请问一下，这个ziggy的仓库源码或者构建发布地址在哪，给linux用的时候看到了这玩意

**Reply by hooke007** [reply]
就是uosc仓库
